File: server/src/main/java/org/dan/jadalnia/app/label/AsyncDao.java
package org.dan.jadalnia.app.label;

import org.jooq.DSLContext;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.function.Function;

import static java.util.concurrent.CompletableFuture.supplyAsync;
import static org.dan.jadalnia.sys.ctx.ExecutorCtx.DEFAULT_EXECUTOR;

public class AsyncDao {
    @Inject
    private DSLContext jooq;
    @Inject
    @Named(DEFAULT_EXECUTOR)
    private ExecutorService executor;

    protected <T> CompletableFuture<T> execQuery(Function<DSLContext, T> query) {
        return supplyAsync(() -> query.apply(jooq), executor);
    }
}

File: service/src/main/kotlin/org/dan/jadalnia/app/label/AsyncDao.kt
package org.dan.jadalnia.app.label

import org.dan.jadalnia.sys.ctx.FutureExecutor
import org.jooq.DSLContext
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.TransactionDefinition
import org.springframework.transaction.support.TransactionTemplate
import java.util.concurrent.CompletableFuture
import javax.inject.Inject

open class AsyncDao {
  @Inject
  protected var tx: PlatformTransactionManager? = null
  @Inject
  protected var jooq: DSLContext? = null
  @Inject
  protected var executor: FutureExecutor? = null

  protected fun <T> execQuery(query: (DSLContext) -> T): CompletableFuture<T> {
    val jq = jooq as DSLContext
    return (executor as FutureExecutor).run {
      val tt = TransactionTemplate(tx!!)
      tt.setIsolationLevel(TransactionDefinition.ISOLATION_READ_COMMITTED)
      tt.execute { query(jq) }
    }
  }
}
